Raise availability guard when core or read replica copy store

lutovich opened the pull request:
PR consists of two commits:

* _Raise availability guard when core or read replica copy store_
Instance might be forced to copy store from another instance if it requires to pull a transaction log that has already been pruned away. This could happen when transaction pulling (catch up) experience problems like network outage. Database is not operational during store copy, most components including the datasource are shutdown so no transactions can be served.
However it was possible for users (bolt or embedded) to attempt start new transactions which would hang because of kernel components being stopped. In similar situations availability guard must be raised to block new transactions.
This commit makes `LocalDatabase` raise availability guard when it is stopped. `LocalDatabase` is the component used by cores and read replicas to manage the lifecycle of the datasource. It is restarted during store copying.

 * _Better error message when db stopped for store copy_
This commit improves error message when users (bolt or embedded) try to start a transaction while cluster member is copying store. It is done using a special `AvailabilityRequirement` for the `AvailabilityGuard` when local database is stopped to perform a store copy.

lutovich committed:
Raise availability guard when core or read replica copy store

Instance might be forced to copy store from another instance if it requires to
pull a transaction log that has already been pruned away. This could happen
when transaction pulling (catch up) experience problems like network outage.
Database is not operational during store copy, most components including the
datasource are shutdown so no transactions can be served.

However it was possible for users (bolt or embedded) to attempt start new
transactions which would hang because of kernel components being stopped.
In similar situations availability guard must be raised to block new
transactions.

This commit makes `LocalDatabase` raise availability guard when it is stopped.
`LocalDatabase` is the component used by cores and read replicas to manage the
lifecycle of the datasource. It is restarted during store copying.

lutovich committed:
Better error message when db stopped for store copy

This commit improves error message when users (bolt or embedded) try to
start a transaction while cluster member is copying store. It is done using a
special `AvailabilityRequirement` for the `AvailabilityGuard` when local
database is stopped to perform a store copy.

davidegrohmann commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/LocalDatabase.java:
```diff
@@ -80,23 +94,29 @@ public synchronized void start() throws Throwable
     public boolean isAvailable()
     {
-        return started;
+        return currentRequirement == null;
```
should be this read through a synchronize?

lutovich replied:
Right, this method should be synchronized. Will fix.

lutovich committed:
Fix field visibility issue in LocalDatabase

It tracks current availability requirement in a field. This field is mostly
accessed from the constructor and lifecycle methods `#start()` and `#stop()`.
Constructor assignment is safe and lifecycle methods are synchronized.

Field is also accessed from `#isAvailable()` method which is not synchronized
and thus it's access was racy. This commit makes field volatile to ensure
correct visibility in `#isAvailable()`.

martinfurmanski commented on enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/LocalDatabase.java:
```diff
@@ -193,4 +213,32 @@ public TransactionCommitProcess getCommitProcess()
+        localCommit = null;
+        dataSourceManager.stop();
+
+        raiseAvailabilityGuard( requirement );
```
Isn't it better to raise this first?

martinfurmanski replied:
I am also concerned that a transaction might still be in progress running, just about to start or already blocking on something before this was raised, but either way this should take care of a a few cases. Perhaps the intention is for internal components to continuously poll on the availability guard for all blocking matters and thus eventually get released. Let's revisit in the future if necessary and perhaps add that if missing somewhere.

lutovich replied:
Yeah, it is definitely better to raise the guard before stopping the datasource. Fixed now.

I think if transaction gets in-between guard and `NeoStoreDataSource#stop()` it'll be asked to terminate during `NSDS#stop()`. Locks are termination aware so transactions should not just block somewhere.

lutovich committed:
Raise availability guard first in `LocalDatabase#stop()`

`LocalDatabase` raises the availability guard when it is stopped. It used to do
it after the datasource manager is stopped which could lead to transactions
starting when datasource is being stopped.

This commit makes `LocalDatabase` raise the guard as the first step
during stop.

martinfurmanski commented:
@lutovich Sorry to be a pain, but just realised we probably want this starting from 3.1.

lutovich commented:
@martinfurmanski I'll open a new PR for 3.1 branch.
